### PR TITLE
Unit test for NSMNSE-37: removeClientInterface() function

### PR DESCRIPTION
**Pull request by ziyuguo716**
PR for [NSMNSE-37](https://jira-eng-sjc4.cisco.com/jira/browse/NSMNSE-37) (Subtask of NSMNSE-18)

Wrote unit test for pkg/unversial-cnf/config/composite.go:: **removeClientInterface()**

- Added `removeClientInterface_test.go` (To be merged into Samuel's composite_test.go)

- Added mock functions in pkg/universal-cnf/config/mocks/UniversalCNFBackend.go

- My test cases cover 100% of lines in **removeClientInterface()** function:

Screen Shot 2020-11-18 at 4 24 50 PM

### CODE REVIEW

File: pkg/universal-cnf/config/removeClientInterface_test.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/cisco-app-networking/nsm-nse/pkg/universal-cnf/config/mocks"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connectioncontext"
+	"go.ligato.io/vpp-agent/v3/proto/ligato/vpp"
+	interfaces "go.ligato.io/vpp-agent/v3/proto/ligato/vpp/interfaces"
```
Comment (ana-moga) [inline pkg/universal-cnf/config/removeClientInterface_test.go]: sort imports

File: pkg/universal-cnf/config/removeClientInterface_test.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+			assert.NotNil(t, err)
+		} else {
+			assert.Nil(t, err)
+			assert.NotNil(t, removeConfig)
+			assert.Equal(t, 1, len(removeConfig.Interfaces))
+		}
+	}
+}
+
+var interfaces1 = []*interfaces.Interface{
```
Comment (ana-moga) [inline pkg/universal-cnf/config/removeClientInterface_test.go]: I would move these variables before the function Test... under one
 ```
var( 
            interfaces1= , 
            interfaces2= ,
)
```

File: pkg/universal-cnf/config/removeClientInterface_test.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+			DstIpAddr:     "192.168.22.3/30",
+			SrcIpRequired: true,
+			DstIpRequired: true,
+			DstRoutes: []*connectioncontext.Route{
+				{Prefix: "192.168.0.0/16"},
+			},
+		},
+	},
+	Labels: map[string]string{
+		"":          "",
```
Comment (ana-moga) [inline pkg/universal-cnf/config/removeClientInterface_test.go]: Why to we need this "": ""? I find it a little bit strange, that's all

File: pkg/universal-cnf/config/removeClientInterface_test.go [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+		"podName":   "helloworld-ucnf-6454c88c5f-pw98x",
+	},
+	Path: &connection.Path{
+		PathSegments: []*connection.PathSegment{
+			{
+				Name: "kind-2-control-plane",
+			},
+		},
+	},
+}
```
Comment (catalinux17) [inline pkg/universal-cnf/config/removeClientInterface_test.go]: add all of this to a `var` block before the function

File: pkg/universal-cnf/config/removeClientInterface_test.go (hunk 2 above)
Comment (ana-moga) [reply pkg/universal-cnf/config/removeClientInterface_test.go]: I would suggest to also change the name of these variables, but not sure into what.

File: pkg/universal-cnf/config/mocks/UniversalCNFBackend.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
+// Code generated by mockery v1.0.0. DO NOT EDIT.
```
Comment (ana-moga) [inline pkg/universal-cnf/config/mocks/UniversalCNFBackend.go:1]: I am not really familiar with the nsm-nse project, but if this is a new tool to generate mocks in this project, consider adding some comments on how to use it. If there is any readme file or smtg.

File: pkg/universal-cnf/config/removeClientInterface_test.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+		removeConfig, err := uce.removeClientInterface(c.conn)
+		if c.expectedError != nil {
+			assert.NotNil(t, err)
+		} else {
+			assert.Nil(t, err)
+			assert.NotNil(t, removeConfig)
+			assert.Equal(t, 1, len(removeConfig.Interfaces))
+		}
+	}
+}
```
Comment (catalinux17) [inline pkg/universal-cnf/config/removeClientInterface_test.go]: You can opt for a more elegant approach. https://play.golang.org/p/XnZoNIDliae. Your call.

**Review by catalinux17 — CHANGES_REQUESTED** [review]

File: pkg/universal-cnf/config/removeClientInterface_test.go (hunk 3 above)
Comment (ziyuguo716) [reply pkg/universal-cnf/config/removeClientInterface_test.go]: Thanks for pointing this out! It was a mistake made by me. Will remove this line

File: pkg/universal-cnf/config/removeClientInterface_test.go (hunk 2 above)
Comment (ziyuguo716) [reply pkg/universal-cnf/config/removeClientInterface_test.go]: Thanks, will do!

File: pkg/universal-cnf/config/mocks/UniversalCNFBackend.go (hunk 5 above)
Comment (ziyuguo716) [reply pkg/universal-cnf/config/mocks/UniversalCNFBackend.go:1]: Thanks for the suggestion! I will add some comments on how to generate mocks

File: pkg/universal-cnf/config/removeClientInterface_test.go (hunk 6 above)
Comment (ziyuguo716) [reply pkg/universal-cnf/config/removeClientInterface_test.go]: Thanks for the review! I will make changes accordingly

File: pkg/universal-cnf/config/removeClientInterface_test.go (hunk 1 above)
Comment (ziyuguo716) [reply pkg/universal-cnf/config/removeClientInterface_test.go]: Hi @AnaMariaMoga , just want to clarify: I believe the imports are already sorted in alphabetical order and are organized into groups. Could you specify how you would like them to be sorted?

Comment (ana-moga) [reply pkg/universal-cnf/config/removeClientInterface_test.go]: Hi, I saw this comment in one PR and I think it clarifies the situation: 
```
Imports should be placed in 3 groups:

- Standard library

- Third-party packages

- First-party packages from outside repositories (e.g. github.com/org/path/to/import)

```

**Review by ana-moga — APPROVED** [review]
Looks good to me with the new changes 👍

**Comment by ziyuguo716** [comment]
Moving this PR to https://github.com/cisco-app-networking/nsm-nse/pull/50

**Pull request closed without merging**